### ISSUE TITLE
Pulling of linkerd container images takes a lot of time in AKS in a fresh cluster

### ISSUE THREAD

**Issue by rohitsakala** [Azure/AKS#2485]
**What happened**:
Linkerd installation fails with timeout error.
I created a Azure Kubernetes Service and ran the following commands

```linkerd install | kubectl apply -f - && linkerd check```

**What you expected to happen**:

Linkerd to be installed correctly. But I face this error from `linkerd check`

```
Warning: resource namespaces/linkerd is missing the kubectl.kubernetes.io/last-applied-configuration annotation which is required by kubectl apply. kubectl apply should only be used on resources created declaratively by either kubectl create --save-config or kubectl apply. The missing annotation will be patched automatically.
namespace/linkerd configured
clusterrole.rbac.authorization.k8s.io/linkerd-linkerd-identity created
clusterrolebinding.rbac.authorization.k8s.io/linkerd-linkerd-identity created
serviceaccount/linkerd-identity created
clusterrole.rbac.authorization.k8s.io/linkerd-linkerd-controller created
clusterrolebinding.rbac.authorization.k8s.io/linkerd-linkerd-controller created
serviceaccount/linkerd-controller created
clusterrole.rbac.authorization.k8s.io/linkerd-linkerd-destination created
clusterrolebinding.rbac.authorization.k8s.io/linkerd-linkerd-destination created
serviceaccount/linkerd-destination created
role.rbac.authorization.k8s.io/linkerd-heartbeat created
rolebinding.rbac.authorization.k8s.io/linkerd-heartbeat created
clusterrole.rbac.authorization.k8s.io/linkerd-heartbeat created
clusterrolebinding.rbac.authorization.k8s.io/linkerd-heartbeat created
serviceaccount/linkerd-heartbeat created
customresourcedefinition.apiextensions.k8s.io/serviceprofiles.linkerd.io created
customresourcedefinition.apiextensions.k8s.io/trafficsplits.split.smi-spec.io created
clusterrole.rbac.authorization.k8s.io/linkerd-linkerd-proxy-injector created
clusterrolebinding.rbac.authorization.k8s.io/linkerd-linkerd-proxy-injector created
serviceaccount/linkerd-proxy-injector created
secret/linkerd-proxy-injector-k8s-tls created
mutatingwebhookconfiguration.admissionregistration.k8s.io/linkerd-proxy-injector-webhook-config created
clusterrole.rbac.authorization.k8s.io/linkerd-linkerd-sp-validator created
clusterrolebinding.rbac.authorization.k8s.io/linkerd-linkerd-sp-validator created
serviceaccount/linkerd-sp-validator created
secret/linkerd-sp-validator-k8s-tls created
validatingwebhookconfiguration.admissionregistration.k8s.io/linkerd-sp-validator-webhook-config created
podsecuritypolicy.policy/linkerd-linkerd-control-plane created
role.rbac.authorization.k8s.io/linkerd-psp created
rolebinding.rbac.authorization.k8s.io/linkerd-psp created
configmap/linkerd-config created
secret/linkerd-identity-issuer created
service/linkerd-identity created
service/linkerd-identity-headless created
deployment.apps/linkerd-identity created
service/linkerd-controller-api created
deployment.apps/linkerd-controller created
service/linkerd-dst created
service/linkerd-dst-headless created
deployment.apps/linkerd-destination created
cronjob.batch/linkerd-heartbeat created
deployment.apps/linkerd-proxy-injector created
service/linkerd-proxy-injector created
service/linkerd-sp-validator created
deployment.apps/linkerd-sp-validator created
secret/linkerd-config-overrides created
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
× controller pod is running
    linkerd-controller-86b587f4d7-4sksh status is PodInitializing
    see https://linkerd.io/2.10/checks/#l5d-existence-controller for hints

Status check results are ×
```

**How to reproduce it (as minimally and precisely as possible)**:

- Create a AKS cluster with 1 node.
- Run this command ```linkerd install | kubectl apply -f - && linkerd check```

**Anything else we need to know?**:

This happens in the fresh cluster only. The pulling of images is taking a lot of time in a fresh cluster. Screenshot attached. The pulling of the image [50 MB] - ```cr.l5d.io/linkerd/controller:stable-2.10.2``` -  takes 5 minutes. that is not good.

![Screenshot 2021-08-10 at 13 55 32](https://user-images.githubusercontent.com/8402361/128863282-c661f492-430d-4af3-a017-908b4480bef9.png)

**Environment**:
- Kubernetes version (use `kubectl version`): 
```
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.0", GitCommit:"cb303e613a121a29364f75cc67d3d580833a7479", GitTreeState:"clean", BuildDate:"2021-04-08T21:16:14Z", GoVersion:"go1.16.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.7", GitCommit:"e1d093448d0ed9b9b1a48f49833ff1ee64c05ba5", GitTreeState:"clean", BuildDate:"2021-06-03T00:20:57Z", GoVersion:"go1.15.12", Compiler:"gc", Platform:"linux/amd64"}
```
- Size of cluster (how many worker nodes are in the cluster?)
1 node - linux - Standard_DS2_v2

**Comment by matucker-msft** [comment]
Huh. I can definitely reproduce this. I am seeing signs that the issue is on the side of cr.l5d.io with both "connection refused" and "connection reset by peer" showing up in my kubelet logs. We might need to engage linkerd here to see what is happening on the registry side itself.

As a possible temporary workaround, we do have all these images hosted in Microsoft Container Registry: mcr.microsoft.com/oss/linkerd/controller:2.10.2 - I am able to successfully pull the images from there very quickly.

**Comment by mboutet** [comment]
Another workaround is to use their images hosted on GitHub's registry at https://github.com/orgs/linkerd/packages.

**Comment by dysinger** [comment]
I reproduced this over and over. It only happens on AKS K8S >1.18. 1.18 with the docker container runtime works fine. As soon as you hit containerd it starts doing this 5 minute delay. 1.19, 1.20, 1.21 all have it. It's pretty consistent if you launch a 5 node cluster. You will see all the pods launch in a few seconds except for one or two that will be 5 minutes and some seconds. Never 3 minutes, Never 4 minutes, Never 6 minutes. Always 5 minutes and some seconds.

**Comment by dysinger** [comment]
As a test I launched a 3 node cluster on GCE using Kubernetes 1.20.x (containerd) in their us-east region. Linkerd launches quickly in a few seconds with no problems. Could this be an issue with Microsoft's fork of containerd?

```
  $ kubectl get events --namespace=linkerd | grep "Successfully pulled"
  113s        Normal   Pulled              pod/linkerd-destination-785744db96-dpt4r       Successfully pulled image "cr.l5d.io/linkerd/proxy-init:v1.3.12" in 3.074303844s
  102s        Normal   Pulled              pod/linkerd-destination-785744db96-dpt4r       Successfully pulled image "cr.l5d.io/linkerd/proxy:edge-21.6.2" in 9.978225328s
  113s        Normal   Pulled              pod/linkerd-identity-7fcf7f556f-x8szd          Successfully pulled image "cr.l5d.io/linkerd/proxy-init:v1.3.12" in 4.001374854s
  108s        Normal   Pulled              pod/linkerd-identity-7fcf7f556f-x8szd          Successfully pulled image "cr.l5d.io/linkerd/controller:edge-21.6.2" in 4.70043899s
  101s        Normal   Pulled              pod/linkerd-identity-7fcf7f556f-x8szd          Successfully pulled image "cr.l5d.io/linkerd/proxy:edge-21.6.2" in 6.851748552s
  111s        Normal   Pulled              pod/linkerd-proxy-injector-75cb8dcb97-lmz9l    Successfully pulled image "cr.l5d.io/linkerd/proxy-init:v1.3.12" in 3.685217792s
  104s        Normal   Pulled              pod/linkerd-proxy-injector-75cb8dcb97-lmz9l    Successfully pulled image "cr.l5d.io/linkerd/proxy:edge-21.6.2" in 6.55855913s
  37s         Normal   Pulled              pod/linkerd-proxy-injector-75cb8dcb97-lmz9l    Successfully pulled image "cr.l5d.io/linkerd/controller:edge-21.6.2" in 4.265147509s
```

**Comment by lupusllc** [comment]
We've been having this issue for a month now. We run a monthly maint script on our clusters which creates a duplicate node pools and kill the old ones. These clusters are running in West US and East US on 1.20.7. Due to these rpc i/o timeouts when pulling images, this work goes from usually taking 15+ min to 90-ish min.

I have been trying to isolate exactly what could be causing these issues. So far I've created 7 test AKS clusters on various versions, 1.20.7, 1.20.9, 1.21.2, with and without Calico, using kubenet and CNI. All had shown the same symptoms, though usually with a fresh cluster, we see a 1-2 min load time without the same errors ( those we have seen them on a few occasions ). But every single time we run our maint script, we see these issues and the rpc timeout i/o errors.

We were previously seeing push issues to an ACR that appeared around this same time frame, but it doesn't seem to be related.

If I run DinD in a pod, on the AKS, I can continually pull down, delete and repeat this process without any issues at all. Something strange is happening in AKS and/or Azure specifically that's causing these issues.

**Comment by aviaviavi** [comment]
Thanks Jorge and Azure team! It appears containerd >= 1.5 solves the issue, available with Kubernetes 1.22+ on AKS.
